File: ios/EscapePhoneTests/ConvenienceStorePuzzleTests.swift
import XCTest
@testable import EscapePhone

final class ConvenienceStorePuzzleTests: XCTestCase {
    // Receipt
    func testSubmitReceiptAnswer_withCorrectAnomaly_succeeds() {
        var engine = ReceiptPuzzleEngine()
        engine.selectReceiptItem(ConvenienceStoreSpec.receiptAnomalyItemId)
        XCTAssertTrue(engine.submitReceiptAnswer())
    }

    func testSubmitReceiptAnswer_withWrongAnomaly_fails() {
        var engine = ReceiptPuzzleEngine()
        engine.selectReceiptItem("water")
        XCTAssertFalse(engine.submitReceiptAnswer())
    }

    func testCompleteReceiptPuzzle_unlocksBarcodePuzzle() {
        var engine = ReceiptPuzzleEngine()
        XCTAssertTrue(engine.completeReceiptPuzzle())
        XCTAssertTrue(engine.isSolved)
    }

    // Barcode
    func testAssembleBarcodeCode_withCorrectSegments_succeeds() {
        var engine = BarcodePuzzleEngine()
        engine.selectBarcodeSegment("category", value: "99")
        engine.selectBarcodeSegment("shelfPosition", value: "40")
        engine.selectBarcodeSegment("arrivalOrder", value: "01")
        XCTAssertEqual(engine.assembleBarcodeCode(), "994001")
        XCTAssertTrue(engine.submitBarcodeRule())
    }

    func testAssembleBarcodeCode_withWrongSegments_fails() {
        var engine = BarcodePuzzleEngine()
        engine.selectBarcodeSegment("category", value: "12")
        engine.selectBarcodeSegment("shelfPosition", value: "05")
        engine.selectBarcodeSegment("arrivalOrder", value: "03")
        XCTAssertFalse(engine.submitBarcodeRule())
    }

    func testCompleteBarcodePuzzle_unlocksShelfDifferencePuzzle() {
        var engine = BarcodePuzzleEngine()
        XCTAssertTrue(engine.completeBarcodePuzzle())
        XCTAssertTrue(engine.isSolved)
    }

    // Shelf difference
    func testSubmitShelfDifferences_withAllTargets_succeeds() {
        var engine = ShelfDifferencePuzzleEngine()
        for target in ConvenienceStoreSpec.shelfDifferenceTargetIds { engine.selectShelfDifference(target) }
        XCTAssertTrue(engine.submitShelfDifferences())
    }

    func testSubmitShelfDifferences_withMissingTarget_fails() {
        var engine = ShelfDifferencePuzzleEngine()
        for target in ConvenienceStoreSpec.shelfDifferenceTargetIds.dropFirst() { engine.selectShelfDifference(target) }
        XCTAssertFalse(engine.submitShelfDifferences())
    }

    func testTiltObjectPuzzle_reachesDestination() {
        var engine = TiltObjectPuzzleEngine(objectX: 0.5, objectY: 0.5)
        engine.applyTiltInput(deltaX: 0, deltaY: 0.4)
        XCTAssertTrue(engine.isSolved)
    }

    func testTouchObjectPuzzle_reachesDestination() {
        var engine = TiltObjectPuzzleEngine()
        engine.applyTouchInput(x: 0.5, y: 0.9)
        XCTAssertTrue(engine.isSolved)
    }

    // CCTV
    func testSubmitCctvSequence_withCorrectOrder_succeeds() {
        var engine = CctvPuzzleEngine()
        let expected = ConvenienceStoreSpec.cctvRecords.sorted { $0.correctOrder < $1.correctOrder }.map(\.id)
        while engine.recordOrder != expected {
            let misplacedIndex = engine.recordOrder.indices.first { engine.recordOrder[$0] != expected[$0] }!
            let targetIndex = engine.recordOrder.firstIndex(of: expected[misplacedIndex])!
            if targetIndex > misplacedIndex { engine.moveCctvRecordUp(targetIndex) } else { engine.moveCctvRecordDown(targetIndex) }
        }
        XCTAssertTrue(engine.submitCctvSequence())
    }

    func testSubmitCctvSequence_withWrongOrder_fails() {
        var engine = CctvPuzzleEngine()
        XCTAssertFalse(engine.submitCctvSequence())
    }

    func testCompleteCctvPuzzle_unlocksInventoryPuzzle() {
        var engine = CctvPuzzleEngine()
        XCTAssertTrue(engine.completeCctvPuzzle())
        XCTAssertTrue(engine.isSolved)
    }

    // Inventory
    func testCalculateExpectedInventory_returnsCorrectValue() {
        let engine = InventoryPuzzleEngine()
        XCTAssertEqual(engine.calculateExpectedInventory("water"), 3)
        XCTAssertEqual(engine.calculateExpectedInventory("evidence_box"), 3)
    }

    func testSubmitInventoryDiscrepancy_withCorrectItem_succeeds() {
        var engine = InventoryPuzzleEngine()
        engine.selectInventoryItem(ConvenienceStoreSpec.discrepancyItemId)
        XCTAssertTrue(engine.submitInventoryDiscrepancy())
    }

    func testSubmitInventoryDiscrepancy_withWrongItem_fails() {
        var engine = InventoryPuzzleEngine()
        engine.selectInventoryItem("water")
        XCTAssertFalse(engine.submitInventoryDiscrepancy())
    }

    // Customer pattern
    func testDecodePurchasePattern_withCorrectSequence_returnsMessage() {
        let engine = CustomerPatternPuzzleEngine()
        XCTAssertEqual(engine.decodePurchasePattern(), ConvenienceStoreSpec.customerPatternMessage)
    }

    func testSubmitCustomerPattern_withWrongSequence_fails() {
        var engine = CustomerPatternPuzzleEngine()
        XCTAssertFalse(engine.submitCustomerPattern("wrong message"))
    }

    func testCompleteCustomerPatternPuzzle_unlocksTimelinePuzzle() {
        var engine = CustomerPatternPuzzleEngine()
        XCTAssertTrue(engine.completeCustomerPatternPuzzle())
        XCTAssertTrue(engine.isSolved)
    }

    // Timeline
    func testValidateIncidentTimeline_withCorrectOrder_succeeds() {
        var engine = IncidentTimelinePuzzleEngine()
        let expected = ConvenienceStoreSpec.incidentEvents.sorted { $0.correctOrder < $1.correctOrder }.map(\.id)
        while engine.eventOrder != expected {
            let misplacedIndex = engine.eventOrder.indices.first { engine.eventOrder[$0] != expected[$0] }!
            engine.moveIncidentEvent(expected[misplacedIndex], newIndex: misplacedIndex)
        }
        XCTAssertTrue(engine.isOrderCorrect())
    }

    func testValidateIncidentTimeline_withWrongOrder_fails() {
        let engine = IncidentTimelinePuzzleEngine()
        XCTAssertFalse(engine.isOrderCorrect())
    }

    func testValidateIncidentTimeline_withCorrectConclusions_succeeds() {
        var engine = IncidentTimelinePuzzleEngine()
        for event in ConvenienceStoreSpec.incidentEvents { engine.selectIncidentConclusion(event.id) }
        XCTAssertTrue(engine.areConclusionsCorrect())
    }
}

File: AOS/app/src/test/java/com/example/escapephone/ConvenienceStorePuzzleTests.kt
package com.example.escapephone

import com.example.escapephone.core.game.BarcodePuzzleEngine
import com.example.escapephone.core.game.CctvPuzzleEngine
import com.example.escapephone.core.game.CustomerPatternPuzzleEngine
import com.example.escapephone.core.game.IncidentTimelinePuzzleEngine
import com.example.escapephone.core.game.InventoryPuzzleEngine
import com.example.escapephone.core.game.ReceiptPuzzleEngine
import com.example.escapephone.core.game.ShelfDifferencePuzzleEngine
import com.example.escapephone.core.game.TiltObjectPuzzleEngine
import com.example.escapephone.core.model.ConvenienceStoreSpec
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ConvenienceStorePuzzleTests {
    // Receipt
    @Test
    fun submitReceiptAnswer_withCorrectAnomaly_succeeds() {
        val engine = ReceiptPuzzleEngine()
        engine.selectReceiptItem(ConvenienceStoreSpec.receiptAnomalyItemId)
        assertTrue(engine.submitReceiptAnswer())
    }

    @Test
    fun submitReceiptAnswer_withWrongAnomaly_fails() {
        val engine = ReceiptPuzzleEngine()
        engine.selectReceiptItem("water")
        assertFalse(engine.submitReceiptAnswer())
    }

    @Test
    fun completeReceiptPuzzle_unlocksBarcodePuzzle() {
        val engine = ReceiptPuzzleEngine()
        assertTrue(engine.completeReceiptPuzzle())
        assertTrue(engine.isSolved)
    }

    // Barcode
    @Test
    fun assembleBarcodeCode_withCorrectSegments_succeeds() {
        val engine = BarcodePuzzleEngine()
        engine.selectBarcodeSegment("category", "99")
        engine.selectBarcodeSegment("shelfPosition", "40")
        engine.selectBarcodeSegment("arrivalOrder", "01")
        assertEquals("994001", engine.assembleBarcodeCode())
        assertTrue(engine.submitBarcodeRule())
    }

    @Test
    fun assembleBarcodeCode_withWrongSegments_fails() {
        val engine = BarcodePuzzleEngine()
        engine.selectBarcodeSegment("category", "12")
        engine.selectBarcodeSegment("shelfPosition", "05")
        engine.selectBarcodeSegment("arrivalOrder", "03")
        assertFalse(engine.submitBarcodeRule())
    }

    @Test
    fun completeBarcodePuzzle_unlocksShelfDifferencePuzzle() {
        val engine = BarcodePuzzleEngine()
        assertTrue(engine.completeBarcodePuzzle())
        assertTrue(engine.isSolved)
    }

    // Shelf difference
    @Test
    fun submitShelfDifferences_withAllTargets_succeeds() {
        val engine = ShelfDifferencePuzzleEngine()
        ConvenienceStoreSpec.shelfDifferenceTargetIds.forEach { engine.selectShelfDifference(it) }
        assertTrue(engine.submitShelfDifferences())
    }

    @Test
    fun submitShelfDifferences_withMissingTarget_fails() {
        val engine = ShelfDifferencePuzzleEngine()
        ConvenienceStoreSpec.shelfDifferenceTargetIds.drop(1).forEach { engine.selectShelfDifference(it) }
        assertFalse(engine.submitShelfDifferences())
    }

    @Test
    fun tiltObjectPuzzle_reachesDestination() {
        val engine = TiltObjectPuzzleEngine(objectX = 0.5f, objectY = 0.5f)
        engine.applyTiltInput(0f, 0.4f)
        assertTrue(engine.isSolved)
    }

    @Test
    fun touchObjectPuzzle_reachesDestination() {
        val engine = TiltObjectPuzzleEngine()
        engine.applyTouchInput(0.5f, 0.9f)
        assertTrue(engine.isSolved)
    }

    // CCTV
    @Test
    fun submitCctvSequence_withCorrectOrder_succeeds() {
        val engine = CctvPuzzleEngine()
        val expected = ConvenienceStoreSpec.cctvRecords.sortedBy { it.correctOrder }.map { it.id }
        while (engine.recordOrder != expected) {
            val misplacedIndex = engine.recordOrder.indices.first { engine.recordOrder[it] != expected[it] }
            val targetIndex = engine.recordOrder.indexOf(expected[misplacedIndex])
            if (targetIndex > misplacedIndex) engine.moveCctvRecordUp(targetIndex) else engine.moveCctvRecordDown(targetIndex)
        }
        assertTrue(engine.submitCctvSequence())
    }

    @Test
    fun submitCctvSequence_withWrongOrder_fails() {
        val engine = CctvPuzzleEngine()
        assertFalse(engine.submitCctvSequence())
    }

    @Test
    fun completeCctvPuzzle_unlocksInventoryPuzzle() {
        val engine = CctvPuzzleEngine()
        assertTrue(engine.completeCctvPuzzle())
        assertTrue(engine.isSolved)
    }

    // Inventory
    @Test
    fun calculateExpectedInventory_returnsCorrectValue() {
        val engine = InventoryPuzzleEngine()
        assertEquals(3, engine.calculateExpectedInventory("water"))
        assertEquals(3, engine.calculateExpectedInventory("evidence_box"))
    }

    @Test
    fun submitInventoryDiscrepancy_withCorrectItem_succeeds() {
        val engine = InventoryPuzzleEngine()
        engine.selectInventoryItem(ConvenienceStoreSpec.discrepancyItemId)
        assertTrue(engine.submitInventoryDiscrepancy())
    }

    @Test
    fun submitInventoryDiscrepancy_withWrongItem_fails() {
        val engine = InventoryPuzzleEngine()
        engine.selectInventoryItem("water")
        assertFalse(engine.submitInventoryDiscrepancy())
    }

    // Customer pattern
    @Test
    fun decodePurchasePattern_withCorrectSequence_returnsMessage() {
        val engine = CustomerPatternPuzzleEngine()
        assertEquals(ConvenienceStoreSpec.customerPatternMessage, engine.decodePurchasePattern())
    }

    @Test
    fun submitCustomerPattern_withWrongSequence_fails() {
        val engine = CustomerPatternPuzzleEngine()
        assertFalse(engine.submitCustomerPattern("wrong message"))
    }

    @Test
    fun completeCustomerPatternPuzzle_unlocksTimelinePuzzle() {
        val engine = CustomerPatternPuzzleEngine()
        assertTrue(engine.completeCustomerPatternPuzzle())
        assertTrue(engine.isSolved)
    }

    // Timeline
    @Test
    fun validateIncidentTimeline_withCorrectOrder_succeeds() {
        val engine = IncidentTimelinePuzzleEngine()
        val expected = ConvenienceStoreSpec.incidentEvents.sortedBy { it.correctOrder }.map { it.id }
        while (engine.eventOrder != expected) {
            val misplacedIndex = engine.eventOrder.indices.first { engine.eventOrder[it] != expected[it] }
            engine.moveIncidentEvent(expected[misplacedIndex], misplacedIndex)
        }
        assertTrue(engine.isOrderCorrect())
    }

    @Test
    fun validateIncidentTimeline_withWrongOrder_fails() {
        val engine = IncidentTimelinePuzzleEngine()
        assertFalse(engine.isOrderCorrect())
    }

    @Test
    fun validateIncidentTimeline_withCorrectConclusions_succeeds() {
        val engine = IncidentTimelinePuzzleEngine()
        ConvenienceStoreSpec.incidentEvents.forEach { engine.selectIncidentConclusion(it.id) }
        assertTrue(engine.areConclusionsCorrect())
    }
}
